src/main/java/frc/robot/commands/TurnToGamepiece.java

```diff
@@ -3,74 +3,49 @@
+import frc.robot.subsystems.Vision;
 import frc.robot.utils.smartshuffleboard.SmartShuffleboard;
 
 public class TurnToGamepiece extends Command {
     private SwerveDrivetrain drivetrain;
-    // private DoubleSupplier x_position;
-    // private DoubleSupplier y_position;
-    private DoubleSubscriber xSub;
-    private DoubleSubscriber ySub;
-    private double x;
-    private double y;
-    ChassisSpeeds driveStates;
-    private boolean finished = false;
-    private final ProfiledPIDController TurningPIDController;
-    private final ProfiledPIDController MovingPIDController;
+    private Vision vision;
+    private double startTime;
+    private double timeSincePieceLoss;
+    private final ProfiledPIDController turningPIDController;
+    private final ProfiledPIDController movingPIDController;
 
-    private double cycle;
-
-    public TurnToGamepiece(SwerveDrivetrain drivetrain) {
+    public TurnToGamepiece(SwerveDrivetrain drivetrain, Vision vision) {
         this.drivetrain = drivetrain;
+        this.vision = vision;
         addRequirements(drivetrain);
-        NetworkTableInstance inst = NetworkTableInstance.getDefault();
-        NetworkTable table = inst.getTable("limelight");
-        xSub = table.getDoubleTopic("tx").subscribe(-1);
-        ySub = table.getDoubleTopic("ty").subscribe(-1);
-
-        TrapezoidProfile.Constraints constraints = new TrapezoidProfile.Constraints(Constants.MAX_ANGULAR_SPEED * 150,
-                2 * Math.PI * 150);
+        TrapezoidProfile.Constraints constraints = new TrapezoidProfile.Constraints(Constants.GAMEPIECE_MAX_VELOCITY, Constants.GAMEPIECE_MAX_ACCELERATION);
         double tP = 0.04;
         double mP = 0.05;
-        TurningPIDController = new ProfiledPIDController(tP, 0, 0, constraints);
-        MovingPIDController = new ProfiledPIDController(mP, 0, 0, constraints);
+        turningPIDController = new ProfiledPIDController(tP, 0, 0, constraints);
+        movingPIDController = new ProfiledPIDController(mP, 0, 0, constraints);
 
     }
 
     @Override
     public void initialize() {
-        cycle = 0;
-        finished = false;
+        startTime = Timer.getFPGATimestamp();
     }
 
     @Override
     public void execute() {
-        x = xSub.get();
-        y = ySub.get();
-        if (y != 0) {
-            driveStates = new ChassisSpeeds(-MovingPIDController.calculate(y + 18), 0,
-                    TurningPIDController.calculate(x));
-            SmartShuffleboard.put("Test", "Speed", TurningPIDController.calculate(x));
-            drivetrain.drive(driveStates);
-            cycle = 0;
-        } else {
-            cycle++;
-        }
-
-        if (y < -17 || cycle > 10) {
-            finished = true;
-        }
+        timeSincePieceLoss = (vision.isPieceSeen() && (Math.abs(vision.getPieceOffestAngleY() - Constants.LIMELIGHT_TURN_TO_PIECE_DESIRED_Y) > Constants.TURN_TO_GAME_PIECE_THRESHOLD)) ? Timer.getFPGATimestamp() : timeSincePieceLoss;
```

Comment:
Add some intermediate values as they seem to be recalculated multiple times

Reply:
Could You elaborate?